File: functions/fn-ingest-jobs/src/adzuna.ts
import { SearchConfig } from "./config";
import { AdzunaJobSearchResults } from "./types/adzuna";

const FETCH_TIMEOUT_MS = 10_000;
const MAX_ATTEMPTS = 5;
const BASE_DELAY_MS = 1_000;
const MAX_DELAY_MS = 16_000;
const RESULTS_PER_PAGE = "50";

export interface AdzunaAuth {
  appId: string;
  appKey: string;
  country: string;
}

export class AdzunaHttpError extends Error {
  constructor(
    public readonly status: number,
    public readonly statusText: string,
    public readonly body: string,
    searchId: string,
  ) {
    super(`Adzuna ${searchId} HTTP ${status} ${statusText}: ${body}`);
    this.name = "AdzunaHttpError";
  }
}

export interface RetryAttemptInfo {
  attempt: number;
  nextDelayMs: number;
  error: string;
}

export type RetryLogger = (info: RetryAttemptInfo) => void;

function buildSearchUrl(auth: AdzunaAuth, search: SearchConfig): URL {
  const url = new URL(
    `https://api.adzuna.com/v1/api/jobs/${auth.country}/search/1`,
  );
  url.searchParams.set("app_id", auth.appId);
  url.searchParams.set("app_key", auth.appKey);
  url.searchParams.set("what", search.what);
  if (search.where) url.searchParams.set("where", search.where);
  url.searchParams.set("results_per_page", RESULTS_PER_PAGE);
  url.searchParams.set("max_days_old", String(search.maxDaysOld ?? 1));
  return url;
}

function isRetryableStatus(status: number): boolean {
  return status >= 500 || status === 429;
}

function isRetryable(err: unknown): boolean {
  if (err instanceof AdzunaHttpError) return isRetryableStatus(err.status);
  if (err instanceof Error) {
    if (err.name === "AbortError" || err.name === "TimeoutError") return true;
    // `fetch failed` is the message Node's undici throws for network errors (DNS, ECONNRESET, etc.)
    if (err.message.includes("fetch failed")) return true;
  }
  return false;
}

function backoffDelay(attempt: number): number {
  const exponential = Math.min(
    BASE_DELAY_MS * 2 ** (attempt - 1),
    MAX_DELAY_MS,
  );
  const jitter = 0.7 + Math.random() * 0.6;
  return Math.round(exponential * jitter);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function fetchAdzunaWithRetry(
  auth: AdzunaAuth,
  search: SearchConfig,
  onRetry?: RetryLogger,
): Promise<AdzunaJobSearchResults> {
  const url = buildSearchUrl(auth, search);

  let lastError: unknown;
  for (let attempt = 1; attempt <= MAX_ATTEMPTS; attempt++) {
    try {
      const response = await fetch(url, {
        signal: AbortSignal.timeout(FETCH_TIMEOUT_MS),
      });
      if (!response.ok) {
        const body = (await response.text()).slice(0, 500);
        throw new AdzunaHttpError(
          response.status,
          response.statusText,
          body,
          search.id,
        );
      }
      return (await response.json()) as AdzunaJobSearchResults;
    } catch (err) {
      lastError = err;
      if (attempt === MAX_ATTEMPTS || !isRetryable(err)) throw err;
      const delay = backoffDelay(attempt);
      onRetry?.({
        attempt,
        nextDelayMs: delay,
        error: err instanceof Error ? err.message : String(err),
      });
      await sleep(delay);
    }
  }
  throw lastError;
}

File: functions/fn-ingest-jobs/adzuna.py
import logging

import requests
from tenacity import (
    before_sleep_log,
    retry,
    retry_if_exception,
    stop_after_attempt,
    wait_exponential_jitter,
)

RESULTS_PER_PAGE = 50
FETCH_TIMEOUT_S = 10


class AdzunaHttpError(Exception):
    def __init__(self, status, body, search_id):
        self.status = status
        self.body = body
        super().__init__(f"Adzuna {search_id} HTTP {status}: {body[:500]}")


def _is_retryable(exc):
    if isinstance(exc, AdzunaHttpError):
        return exc.status >= 500 or exc.status == 429
    return isinstance(exc, (requests.ConnectionError, requests.Timeout))


@retry(
    stop=stop_after_attempt(5),
    wait=wait_exponential_jitter(initial=1, max=16, jitter=2),
    retry=retry_if_exception(_is_retryable),
    reraise=True,
    before_sleep=before_sleep_log(logging.getLogger(__name__), logging.WARNING),
)
def fetch_adzuna(app_id, app_key, country, search):
    params = {
        "app_id": app_id,
        "app_key": app_key,
        "what": search["what"],
        "results_per_page": RESULTS_PER_PAGE,
        "max_days_old": search.get("maxDaysOld", 1),
    }
    if search.get("where"):
        params["where"] = search["where"]

    resp = requests.get(
        f"https://api.adzuna.com/v1/api/jobs/{country}/search/1",
        params=params,
        timeout=FETCH_TIMEOUT_S,
    )
    if not resp.ok:
        raise AdzunaHttpError(resp.status_code, resp.text, search["id"])

    return resp.json()
